geode-wan/src/main/java/org/apache/geode/management/internal/cli/functions/WanCopyRegionFunction.java

```diff
@@ -0,0 +1,581 @@
+import org.apache.geode.internal.cache.wan.InternalGatewaySender;
+import org.apache.geode.internal.serialization.KnownVersion;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.management.cli.CliFunction;
+import org.apache.geode.management.internal.functions.CliFunctionResult;
+import org.apache.geode.management.internal.i18n.CliStrings;
+
+/**
+ * Class for copying via WAN the contents of a region
+ * It must be executed in all members of the Geode cluster that host the region
+ * to be copied. (called with onServers() or withMembers() passing the list
+ * of all members hosting the region).
+ * It also offers the possibility to cancel an ongoing execution of this function.
+ * The copying itself is executed in a new thread with a known name
+ * (parameterized with the regionName and senderId) in order to allow
+ * to cancel ongoing invocations by interrupting that thread.
+ *
+ * It accepts the following arguments in an array of objects
+ * 0: regionName (String)
+ * 1: senderId (String)
+ * 2: isCancel (Boolean): If true, it indicates that an ongoing execution of this
+ * function for the given region and senderId must be stopped. Otherwise,
+ * it indicates that the region must be copied.
+ * 3: maxRate (Long) maximum copy rate in entries per second. In the case of
+ * parallel gateway senders, the maxRate is per server hosting the region.
+ * 4: batchSize (Integer): the size of the batches. Region entries are copied in batches of the
+ * passed size. After each batch is sent, the function checks if the command
+ * must be canceled and also sleeps for some time if necessary to adjust the
+ * copy rate to the one passed as argument.
+ */
+public class WanCopyRegionFunction extends CliFunction<Object[]> implements Declarable {
+  private static final Logger logger = LogService.getLogger();
+  private static final long serialVersionUID = 1L;
+
+  public static final String ID = WanCopyRegionFunction.class.getName();
+
+  private static final int MAX_BATCH_SEND_RETRIES = 1;
+
+  private final Clock clock;
+  private final ThreadSleeper threadSleeper;
+  private final WanCopyRegionFunctionExecutorFactory executorFactory;
+  private final EventCreator eventCreator;
+
+  /**
+   * Contains the ongoing executions of this function
+   */
+  private static final Map<String, Future<CliFunctionResult>> executions =
+      new ConcurrentHashMap<>();
+
+  private int batchId = 0;
+
+  public WanCopyRegionFunction() {
+    this(Clock.systemDefaultZone(), new ThreadSleeper(),
+        new WanCopyRegionFunctionExecutorFactoryImpl(), new EventCreatorImpl());
+  }
+
+  @VisibleForTesting
+  WanCopyRegionFunction(Clock clock, ThreadSleeper threadSleeper,
+      WanCopyRegionFunctionExecutorFactory executorFactory, EventCreator eventCreator) {
+    this.clock = clock;
+    this.threadSleeper = threadSleeper;
+    this.executorFactory = executorFactory;
+    this.eventCreator = eventCreator;
+  }
```

Comment:
I like the change but there were a problems when deserializing the two lambdas passed. So, in order to avoid them, I had to create a ThreadSleeperImpl and a WanCopyRegionFuncionServiceProviderImpl classes with the above's labmdas behaviors.